### PR TITLE
Add kwargs to arguments for Backend.get_compiled_circuit

### PR DESCRIPTION
**Pull request by sjdilkes**

### CODE REVIEW

File: pytket/pytket/backends/backend.py [hunk 1]
```diff
@@ -172,18 +173,19 @@ def default_compilation_pass(self, optimisation_level: int = 2) -> BasePass:
...
         """
-        Return a single circuit compiled with :py:meth:`default_compilation_pass`. See
-        :py:meth:`Backend.get_compiled_circuits`.
+        Return a single circuit compiled with :py:meth:`default_compilation_pass` See
+        :py:meth:`Backend.get_compiled_circuits`. Valid kwargs are Backend specific.
+
         """
         return_circuit = circuit.copy()
-        self.default_compilation_pass(optimisation_level).apply(return_circuit)
+        self.default_compilation_pass(optimisation_level, kwargs).apply(return_circuit)
```
Comment (cqc-alec) [inline pytket/pytket/backends/backend.py:184]:
```suggestion
        self.default_compilation_pass(optimisation_level, **kwargs).apply(return_circuit)
```

**Review by cqc-alec — CHANGES_REQUESTED** [review]

File: pytket/pytket/backends/backend.py [hunk 2]
```diff
@@ -144,7 +143,9 @@ def rebase_pass(self) -> BasePass:
         ...
 
     @abstractmethod
-    def default_compilation_pass(self, optimisation_level: int = 2) -> BasePass:
+    def default_compilation_pass(
+        self, optimisation_level: int = 2, **kwarfs
```
[the remark this answers is not in the document]
Comment (sjdilkes) [reply pytket/pytket/backends/backend.py:147]: Now that's an embarrassing one, fixed!

File: pytket/pytket/backends/backend.py [hunk 3]
```diff
@@ -213,7 +217,9 @@ def get_compiled_circuits(
         :return: Compiled circuits.
         :rtype: List[Circuit]
         """
-        return [self.get_compiled_circuit(c, optimisation_level) for c in circuits]
+        return [
+            self.get_compiled_circuit(c, optimisation_level, kwargs) for c in circuits
```
[the remark this answers is not in the document]
Comment (sjdilkes) [reply pytket/pytket/backends/backend.py:221]: done

File: pytket/pytket/backends/backend.py (hunk 1 above)
Comment (sjdilkes) [reply pytket/pytket/backends/backend.py:184]: done

**Review by cqc-alec — COMMENTED** [review]
Ah! Is this going to break all the backends?

**Comment by sjdilkes** [comment]
> Ah! Is this going to break all the backends?

Yeah, they're all going need to have `**kwargs` added to their corresponding `default_compilation_pass` methods. I think it's worth it to allow users to pass backend specific optimisation choices through the usual interface, and at least any existing user code will work fine.

**Comment by cqc-alec** [comment]
> > Ah! Is this going to break all the backends?
> 
> Yeah, they're all going need to have `**kwargs` added to their corresponding `default_compilation_pass` methods. I think it's worth it to allow users to pass backend specific optimisation choices through the usual interface, and at least any existing user code will work fine.

But it does mean that if someone updates pytket their code will stop working, which breaks the 1.x API promise.

We could make new `Backend` methods `get_default_compilation_pass_v2`, `get_compiled_circuit_v2` and `get_compiled_circuits_v2` which take the kwargs and call the existing methods by default, and override them in pytket-quantinuum (making the old methods call the new ones)?

**Comment by sjdilkes** [comment]
> > > Ah! Is this going to break all the backends?
> > 
> > 
> > Yeah, they're all going need to have `**kwargs` added to their corresponding `default_compilation_pass` methods. I think it's worth it to allow users to pass backend specific optimisation choices through the usual interface, and at least any existing user code will work fine.
> 
> But it does mean that if someone updates pytket their code will stop working, which breaks the 1.x API promise.
> 
> We could make new `Backend` methods `get_default_compilation_pass_v2`, `get_compiled_circuit_v2` and `get_compiled_circuits_v2` which take the kwargs and call the existing methods by default, and override them in pytket-quantinuum (making the old methods call the new ones)?

Good point, I didn't think about people updating `pytket` but not the corresponding extension package.

That's a good idea - I'll do that now.

**Comment by sjdilkes** [comment]
> > > > Ah! Is this going to break all the backends?
> > > 
> > > 
> > > Yeah, they're all going need to have `**kwargs` added to their corresponding `default_compilation_pass` methods. I think it's worth it to allow users to pass backend specific optimisation choices through the usual interface, and at least any existing user code will work fine.
> > 
> > 
> > But it does mean that if someone updates pytket their code will stop working, which breaks the 1.x API promise.
> > We could make new `Backend` methods `get_default_compilation_pass_v2`, `get_compiled_circuit_v2` and `get_compiled_circuits_v2` which take the kwargs and call the existing methods by default, and override them in pytket-quantinuum (making the old methods call the new ones)?
> 
> Good point, I didn't think about people updating `pytket` but not the corresponding extension package.
> 
> That's a good idea - I'll do that now.

Do you think I should also include `optimisation_level` in the kwargs?

**Comment by cqc-alec** [comment]
Rethinking this whole thing...

**Pull request closed without merging**